internal/command/show.go

```diff
@@ -149,42 +174,129 @@ func (c *ShowCommand) showFromLatestStateSnapshot() (*statefile.File, tfdiags.Di
 	return stateFile, diags
 }
 
-func (c *ShowCommand) showFromPath(path string) (*plans.Plan, *statefile.File, *configs.Config, tfdiags.Diagnostics) {
+func (c *ShowCommand) showFromPath(path string) (*plans.Plan, *cloudplan.RemotePlanJSON, *statefile.File, *configs.Config, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 	var planErr, stateErr error
 	var plan *plans.Plan
+	var jsonPlan *cloudplan.RemotePlanJSON
 	var stateFile *statefile.File
 	var config *configs.Config
 
-	// Try to get the plan file and associated data from
-	// the path argument. If that fails, try to get the
-	// statefile from the path argument.
-	plan, stateFile, config, planErr = getPlanFromPath(path)
+	// Path might be a local plan file, a bookmark to a saved cloud plan, or a
+	// state file. First, try to get a plan and associated data from a local
+	// plan file. If that fails, try to get a json plan from the path argument.
+	// If that fails, try to get the statefile from the path argument.
+	plan, jsonPlan, stateFile, config, planErr = c.getPlanFromPath(path)
 	if planErr != nil {
 		stateFile, stateErr = getStateFromPath(path)
 		if stateErr != nil {
-			diags = diags.Append(
-				tfdiags.Sourceless(
-					tfdiags.Error,
-					"Failed to read the given file as a state or plan file",
-					fmt.Sprintf("State read error: %s\n\nPlan read error: %s", stateErr, planErr),
-				),
-			)
-			return nil, nil, nil, diags
+			// To avoid spamming the user with irrelevant errors, first check to
+			// see if one of our errors happens to know for a fact what file
+			// type we were dealing with. If so, then we can ignore the other
+			// ones (which are likely to be something unhelpful like "not a
+			// valid zip file"). If not, we can fall back to dumping whatever
+			// we've got.
+			var unLocal *planfile.ErrUnusableLocalPlan
+			var unState *statefile.ErrUnusableState
+			var unMisc *errUnusableDataMisc
```

Comment:
I appreciate the attention to detail on the error handling here, and also these variable names 😆